### PR TITLE
Add topic model

### PR DESCRIPTION
**Pull request by jackrootstrap**
## Topic model

* [Trello Card #16 As a user I should be able to see all the topics in the system](https://trello.com/c/DFgFO8vV/33-1-as-a-user-i-should-be-able-to-see-all-the-topics-in-the-system)
---
#### Description:

* Adding topic model.
* Enable active storage.
* Add shoulda-matchers gem

### CODE REVIEW

File: db/migrate/20220928175746_create_topics.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class CreateTopics < ActiveRecord::Migration[7.0]
+  def change
+    create_table :topics do |t|
+      t.string :name
```
Comment (JulianPasquale) [inline db/migrate/20220928175746_create_topics.rb]: We should add a not null constraints here. Maybe a unique index too? 🤔

Comment (JulianPasquale) [reply db/migrate/20220928175746_create_topics.rb]: And please add the same validations to the model

Comment (jackrootstrap) [reply db/migrate/20220928175746_create_topics.rb]: Sure, I will add those validations! Thanks! @JulianPasquale

File: app/models/topic.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+#  updated_at :datetime         not null
+#
+# Indexes
+#
+#  index_topics_on_name  (name) UNIQUE
+#
+class Topic < ApplicationRecord
+  has_one_attached :image
+
+  validates :name, uniqueness: true
```
Comment (JulianPasquale) [inline app/models/topic.rb]:
```suggestion
  validates :name, presence: true, uniqueness: true
```

File: spec/factories/topics.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#  name       :string           not null
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+# Indexes
+#
+#  index_topics_on_name  (name) UNIQUE
+#
+FactoryBot.define do
+  factory :topic do
```
Comment (JulianPasquale) [inline spec/factories/topics.rb:17]: According to the FactoryBot best practices:

> It is recommended that you have one factory for each class that provides the simplest set of attributes necessary to create an instance of that class

https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#best-practices

I'm wondering if the image is actualy part of "the simplest set of attributes".

Maybe we can do something like this:
```ruby
FactoryBot.define do
  factory :topic do
    name { Faker::Lorem.word }

    trait :with_image do
      transient do
        sequence(:tmp_img) { |n| Tempfile.new(["tmp_img_#{n}", 'jpg']).path }
      end

      image { Rack::Test::UploadedFile.new(tmp_img, 'image/png') }
    end
  end
end
```

So then
```ruby
let(:topic) { create(:topic) } # It only has a name

let(:topic) { create(:topic, :with_image) } # It has name and image
```

**Review by JulianPasquale — APPROVED** [review]

File: spec/factories/topics.rb (hunk 3 above)
Comment (jackrootstrap) [reply spec/factories/topics.rb:17]: Good one, I will update this factory! Thanks! @JulianPasquale

File: spec/models/topic_spec.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  describe 'factory' do
+    it { is_expected.to be_valid }
+  end
+
+  describe 'attributes' do
+    it { is_expected.to have_db_column(:name) }
+  end
+
+  describe 'validations' do
+    it { is_expected.to validate_uniqueness_of(:name) }
```
Comment (JulianPasquale) [inline spec/models/topic_spec.rb:30]: Let's add a test for the `presence` validation as well

**Review by JulianPasquale — APPROVED** [review]

**Review by sebastiancaraballo — APPROVED** [review]

**Pull request merged**